src/ui/src/shape_list_widget.cpp

```diff
@@ -1,10 +1,172 @@
+}
+
+QVariant TreeModel::headerData(int section, Qt::Orientation orientation,
+                               int role) const {
+	if (orientation == Qt::Horizontal && role == Qt::DisplayRole)
+		return rootItem->data(section);
+	return QVariant();
+}
+
+QModelIndex TreeModel::index(int row, int column,
+                             const QModelIndex& parent) const {
+	if (!hasIndex(row, column, parent))
+		return QModelIndex();
+	TreeItem* parentItem;
+	if (!parent.isValid())
+		parentItem = rootItem;
+	else
+		parentItem = static_cast<TreeItem*>(parent.internalPointer());
+	TreeItem* childItem = parentItem->child(row);
+	if (childItem)
+		return createIndex(row, column, childItem);
+	return QModelIndex();
+}
+
+QModelIndex TreeModel::parent(const QModelIndex& index) const {
+	if (!index.isValid())
+		return QModelIndex();
+	TreeItem* childItem = static_cast<TreeItem*>(index.internalPointer());
+	TreeItem* parentItem = childItem->parentItem();
+	if (parentItem == rootItem)
+		return QModelIndex();
+	return createIndex(parentItem->row(), 0, parentItem);
+}
+
+int TreeModel::rowCount(const QModelIndex& parent) const {
+	if (parent.column() > 0)
+		return 0;
+	TreeItem* parentItem;
+	if (!parent.isValid())
+		parentItem = rootItem;
+	else
+		parentItem = static_cast<TreeItem*>(parent.internalPointer());
+
+	return parentItem->child_count();
+}
+
+void TreeModel::TestFillWithTxtFile(const QString& file_path) {
+	QFile file(file_path);
+	file.open(QIODevice::ReadOnly);
+	QString raw_data = file.readAll();
+	QStringList lines = raw_data.split('\n');
+	file.close();
+
+
+	QVector<TreeItem*> parents;
+	QVector<int> indentations;
+	parents << rootItem;
+	indentations << 0;
+
+	int number = 0;
+
+	while (number < lines.count()) {
+		int position = 0;
+		while (position < lines[number].length()) {
+			if (lines[number].at(position) != ' ')
+				break;
+			position++;
+		}
+		const QString lineData = lines[number].mid(position).trimmed();
+		if (!lineData.isEmpty()) {
+			// Read the column data from the rest of the line.
+			const QStringList columnStrings =
+					lineData.split(QLatin1Char('\t'), Qt::SkipEmptyParts);
+			QVector<QVariant> columnData;
+			columnData.reserve(columnStrings.count());
+			for (const QString& columnString : columnStrings)
+				columnData << columnString;
+
+			if (position > indentations.last()) {
+				// The last child of the current parent is now the new parent
+				// unless the current parent has no children.
+
+				if (parents.last()->child_count() > 0) {
+					parents << parents.last()->child(parents.last()->child_count() - 1);
+					indentations << position;
+				}
+			} else {
+				while (position < indentations.last() && parents.count() > 0) {
+					parents.pop_back();
+					indentations.pop_back();
+				}
+			}
+			// Append a new item to the current parent's list of children.
+			parents.last()->append_child(new TreeItem(columnData, parents.last()));
+		}
+		++number;
+	}
```

Comment:
needed for tests only, we should provide proper implementation later

Reply:
For future PR